diff --git a/javamelody-core/src/main/java/net/bull/javamelody/HtmlCoreReport.java b/javamelody-core/src/main/java/net/bull/javamelody/HtmlCoreReport.java
index 753b7976b..b99138973 100644
--- a/javamelody-core/src/main/java/net/bull/javamelody/HtmlCoreReport.java
+++ b/javamelody-core/src/main/java/net/bull/javamelody/HtmlCoreReport.java
@@ -93,7 +93,8 @@ void writeCustomPeriodLink(Range range, String graphName, String part) throws IO
 			writeln("<input type='hidden' name='period' value=''/>");
 			if (graphName != null) {
 				writeln("<input type='hidden' name='part' value='" + part + "'/>");
-				writeln("<input type='hidden' name='graph' value='" + graphName + "'/>");
+				writeln("<input type='hidden' name='graph' value='" + I18N.urlEncode(graphName)
+						+ "'/>");
 			}
 			writeln("</form><br/>");
 			writeln(END_DIV);
@@ -453,7 +454,8 @@ void writeCounterSummaryPerClass(String counterName, String requestId) throws IO
 		final String separator = "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ";
 		writeln(separator);
 		final String hrefStart = "<a href='?part=counterSummaryPerClass&amp;counter="
-				+ counter.getName() + (requestId == null ? "" : "&amp;graph=" + requestId);
+				+ counter.getName()
+				+ (requestId == null ? "" : "&amp;graph=" + I18N.urlEncode(requestId));
 		writeln(hrefStart + "'>");
 		writeln("<img src='?resource=action_refresh.png' alt='#Actualiser#'/> #Actualiser#</a>");
 
@@ -715,7 +717,8 @@ void writeRefreshAndPeriodLinks(String graphName, String part) throws IOExceptio
 			writeln(separator);
 			writeln("<a href='?'><img src='?resource=action_home.png' alt='#Page_principale#'/> #Page_principale#</a>");
 			writeln(separator);
-			write("<a href='?part=" + part + graphParameter + graphName + "' title='#Rafraichir#'>");
+			write("<a href='?part=" + part + graphParameter + I18N.urlEncode(graphName)
+					+ "' title='#Rafraichir#'>");
 		}
 		write("<img src='?resource=action_refresh.png' alt='#Actualiser#'/> #Actualiser#</a>");
 		if (graphName == null && PDF_ENABLED) {
@@ -735,8 +738,8 @@ void writeRefreshAndPeriodLinks(String graphName, String part) throws IOExceptio
 			if (graphName == null) {
 				write("<a href='?period=" + myPeriod.getCode() + "' ");
 			} else {
-				write("<a href='?part=" + part + graphParameter + graphName + "&amp;period="
-						+ myPeriod.getCode() + "' ");
+				write("<a href='?part=" + part + graphParameter + I18N.urlEncode(graphName)
+						+ "&amp;period=" + myPeriod.getCode() + "' ");
 			}
 			write("title='" + I18N.getFormattedString("Choisir_periode", myPeriod.getLinkLabel())
 					+ "'>");
diff --git a/javamelody-core/src/main/java/net/bull/javamelody/HtmlCounterReport.java b/javamelody-core/src/main/java/net/bull/javamelody/HtmlCounterReport.java
index cba077158..9a1d36e96 100644
--- a/javamelody-core/src/main/java/net/bull/javamelody/HtmlCounterReport.java
+++ b/javamelody-core/src/main/java/net/bull/javamelody/HtmlCounterReport.java
@@ -108,9 +108,10 @@ void writeRequestAndGraphDetail(Collector collector, CollectorServer collectorSe
 				writeln("</div></div>");
 
 				writeln("<div align='center'><img class='synthèse' id='img' src='"
-						+ "?width=960&amp;height=400&amp;graph=" + graphName
+						+ "?width=960&amp;height=400&amp;graph=" + I18N.urlEncode(graphName)
 						+ "' alt='zoom'/></div>");
-				writeln("<div align='right'><a href='?part=lastValue&amp;graph=" + graphName
+				writeln("<div align='right'><a href='?part=lastValue&amp;graph="
+						+ I18N.urlEncode(graphName)
 						+ "' title=\"#Lien_derniere_valeur#\">_</a></div>");
 
 				writeGraphDetailScript(graphName);
@@ -387,7 +388,7 @@ private void writeGraphDetailScript(String graphName) throws IOException {
 			writeln("  height = Math.round(width * initialHeight / initialWidth) - 48;");
 			// reload the images
 			// rq : on utilise des caractères unicode pour éviter des warnings
-			writeln("  document.getElementById('img').src = '?graph=" + graphName
+			writeln("  document.getElementById('img').src = '?graph=" + I18N.urlEncode(graphName)
 					+ "\\u0026width=' + width + '\\u0026height=' + height;");
 			writeln("  document.getElementById('img').style.width = '';");
 			writeln("}");
diff --git a/javamelody-core/src/main/java/net/bull/javamelody/HtmlSessionInformationsReport.java b/javamelody-core/src/main/java/net/bull/javamelody/HtmlSessionInformationsReport.java
index b6d1ffe01..80067f069 100644
--- a/javamelody-core/src/main/java/net/bull/javamelody/HtmlSessionInformationsReport.java
+++ b/javamelody-core/src/main/java/net/bull/javamelody/HtmlSessionInformationsReport.java
@@ -133,7 +133,7 @@ private void writeBackAndRefreshLinks() throws IOException {
 	private void writeBackAndRefreshLinksForSession(String sessionId) throws IOException {
 		writeln("<div class='noPrint'>");
 		writeln("<a href='javascript:history.back()'><img src='?resource=action_back.png' alt='#Retour#'/> #Retour#</a>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;");
-		writeln(A_HREF_PART_SESSIONS + "&amp;sessionId=" + sessionId + "'>");
+		writeln(A_HREF_PART_SESSIONS + "&amp;sessionId=" + I18N.urlEncode(sessionId) + "'>");
 		writeln("<img src='?resource=action_refresh.png' alt='#Actualiser#'/> #Actualiser#</a>");
 		writeln("</div>");
 	}
@@ -142,9 +142,9 @@ private void writeSession(SessionInformations session, boolean displayUser) thro
 		final String nextColumnAlignRight = "</td><td align='right'>";
 		final String nextColumnAlignCenter = "</td><td align='center'>";
 		write("<td><a href='?part=sessions&amp;sessionId=");
-		write(session.getId());
+		write(htmlEncode(session.getId()));
 		write("'>");
-		write(session.getId());
+		write(htmlEncode(session.getId()));
 		write("</a>");
 		write(nextColumnAlignRight);
 		write(durationFormat.format(session.getLastAccess()));
@@ -185,7 +185,7 @@ private void writeSession(SessionInformations session, boolean displayUser) thro
 		write("</td><td align='center' class='noPrint'>");
 		write(A_HREF_PART_SESSIONS);
 		write("&amp;action=invalidate_session&amp;sessionId=");
-		write(session.getId());
+		write(I18N.urlEncode(session.getId()));
 		write("' onclick=\"javascript:return confirm('"
 				+ I18N.getStringForJavascript("confirm_invalidate_session") + "');\">");
 		write("<img width='16' height='16' src='?resource=user-trash.png' alt='#invalidate_session#' title='#invalidate_session#' />");
@@ -219,11 +219,11 @@ void writeSessionDetails(String sessionId, SessionInformations sessionInformatio
 		writeln("<br/>");
 
 		if (sessionInformations == null) {
-			writeln(I18N.getFormattedString("session_invalidee", sessionId));
+			writeln(I18N.getFormattedString("session_invalidee", htmlEncode(sessionId)));
 			return;
 		}
 		writeln("<img width='24' height='24' src='?resource=system-users.png' alt='#Sessions#' />&nbsp;");
-		writeln("<b>" + I18N.getFormattedString("Details_session", sessionId) + "</b>");
+		writeln("<b>" + I18N.getFormattedString("Details_session", htmlEncode(sessionId)) + "</b>");
 		writeSessions(Collections.singletonList(sessionInformations));
 
 		writeln("<br/><b>#Attributs#</b>");
diff --git a/javamelody-core/src/main/java/net/bull/javamelody/I18N.java b/javamelody-core/src/main/java/net/bull/javamelody/I18N.java
index efe9a2ed3..54fff4f66 100644
--- a/javamelody-core/src/main/java/net/bull/javamelody/I18N.java
+++ b/javamelody-core/src/main/java/net/bull/javamelody/I18N.java
@@ -134,6 +134,10 @@ static String javascriptEncode(String text) {
 				.replace("'", "\\'");
 	}
 
+	static String urlEncode(String text) {
+		return javascriptEncode(text);
+	}
+
 	/**
 	 * Encode pour affichage en html.
 	 * @param text message à encoder
