diff --git a/src/main/java/org/primefaces/application/DialogNavigationHandler.java b/src/main/java/org/primefaces/application/DialogNavigationHandler.java
index 56d644a5bb..b694a0c484 100644
--- a/src/main/java/org/primefaces/application/DialogNavigationHandler.java
+++ b/src/main/java/org/primefaces/application/DialogNavigationHandler.java
@@ -25,6 +25,7 @@
 import javax.faces.application.NavigationCase;
 import javax.faces.context.FacesContext;
 import org.primefaces.PrimeFaces;
+import org.primefaces.util.ComponentUtils;
 import org.primefaces.util.Constants;
 
 public class DialogNavigationHandler extends ConfigurableNavigationHandler {
@@ -57,6 +58,7 @@ public void handleNavigation(FacesContext context, String fromAction, String out
             }
 
             String url = context.getApplication().getViewHandler().getBookmarkableURL(context, toViewId, params, includeViewParams);
+            url = ComponentUtils.escapeEcmaScriptText(url);
 
             StringBuilder sb = new StringBuilder();
             String sourceComponentId = (String) attrs.get(Constants.DIALOG_FRAMEWORK.SOURCE_COMPONENT);
@@ -65,6 +67,7 @@ public void handleNavigation(FacesContext context, String fromAction, String out
             if (pfdlgcid == null) {
                 pfdlgcid = UUID.randomUUID().toString();
             }
+            pfdlgcid = ComponentUtils.escapeEcmaScriptText(pfdlgcid);
 
             sb.append("PrimeFaces.openDialog({url:'").append(url).append("',pfdlgcid:'").append(pfdlgcid)
                     .append("',sourceComponentId:'").append(sourceComponentId).append("'");
