diff --git a/nd4j/nd4j-common/src/main/java/org/nd4j/util/ArchiveUtils.java b/nd4j/nd4j-common/src/main/java/org/nd4j/util/ArchiveUtils.java
index 24e2432b54e..70bee869d48 100644
--- a/nd4j/nd4j-common/src/main/java/org/nd4j/util/ArchiveUtils.java
+++ b/nd4j/nd4j-common/src/main/java/org/nd4j/util/ArchiveUtils.java
@@ -65,7 +65,15 @@ public static void unzipFileTo(String file, String dest) throws IOException {
 
                 while (ze != null) {
                     String fileName = ze.getName();
+
+                    String canonicalDestinationDirPath = new File(dest).getCanonicalPath();
                     File newFile = new File(dest + File.separator + fileName);
+                    String canonicalDestinationFile = newFile.getCanonicalPath();
+
+                    if (!canonicalDestinationFile.startsWith(canonicalDestinationDirPath + File.separator)) {
+                        log.debug("Attempt to unzip entry is outside of the target dir");
+                        throw new IOException("Entry is outside of the target dir: ");
+                    }
 
                     if (ze.isDirectory()) {
                         newFile.mkdirs();
