diff --git a/pdfbox/src/test/java/org/apache/pdfbox/encryption/TestSymmetricKeyEncryption.java b/pdfbox/src/test/java/org/apache/pdfbox/encryption/TestSymmetricKeyEncryption.java
index 6a6094852c..b9a30ef09b 100644
--- a/pdfbox/src/test/java/org/apache/pdfbox/encryption/TestSymmetricKeyEncryption.java
+++ b/pdfbox/src/test/java/org/apache/pdfbox/encryption/TestSymmetricKeyEncryption.java
@@ -168,7 +168,10 @@ private void checkPerms(byte[] inputFileAsByteArray, String password,
 
         // check permissions
         assertEquals(expectedPermissions.isOwnerPermission(), currentAccessPermission.isOwnerPermission());
-        assertEquals(expectedPermissions.isReadOnly(), currentAccessPermission.isReadOnly());
+        if (!expectedPermissions.isOwnerPermission())
+        {
+            assertEquals(true, currentAccessPermission.isReadOnly());
+        }
         assertEquals(expectedPermissions.canAssembleDocument(), currentAccessPermission.canAssembleDocument());
         assertEquals(expectedPermissions.canExtractContent(), currentAccessPermission.canExtractContent());
         assertEquals(expectedPermissions.canExtractForAccessibility(), currentAccessPermission.canExtractForAccessibility());
